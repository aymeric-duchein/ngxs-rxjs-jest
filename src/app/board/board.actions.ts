import {Issue, IssueStatus} from '../core/issue/issue.model';

export class BoardIssueClick {
  static readonly type = '[Board] Issue click';

  constructor(public issue: Issue) {
  }
}

export class BoardIssueDropped {
  static readonly type = '[Board] Issue dropped';

  constructor(public droppedIssue: Issue, public status: IssueStatus) {
  }
}
