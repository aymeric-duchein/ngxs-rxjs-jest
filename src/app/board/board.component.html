<div cdkDropListGroup class="board-list-container">
  <app-list class="mat-elevation-z2" [issues]="todoIssues$ | async" [status]="IssueStatus.TODO"></app-list>
  <app-list class="mat-elevation-z2" [issues]="inPRIssues$ | async" [status]="IssueStatus.IN_PROGRESS"></app-list>
  <app-list class="mat-elevation-z2" [issues]="doneIssues$ | async" [status]="IssueStatus.DONE"></app-list>
</div>

<app-add></app-add>

