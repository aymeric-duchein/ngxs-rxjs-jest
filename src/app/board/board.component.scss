@import '~@angular/material/theming';

@include mat-core();

app-add {
  position: absolute;
  bottom: 1rem;
  right: 3rem;
}

.board-list-container {
  height: calc(100vh - 104px);
  padding: 20px 10px;
  color: rgba(0, 0, 0, 0.87);
  display: flex;
  flex-direction: row;
  align-items: stretch;
  justify-content: space-between;
  font-size: 14px;
  position: relative;

  app-list {
    min-width: 30vw;
    max-width: 30vw;
    height: inherit;
    background: #fafafa;
    border-radius: 10px;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
  }
}

