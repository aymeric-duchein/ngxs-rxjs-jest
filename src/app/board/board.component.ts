import {Component} from '@angular/core';
import {Issue, IssueStatus} from '../core/issue/issue.model';
import {IssueState} from '../core/issue/issue.state';
import {Observable} from 'rxjs';
import {Select} from '@ngxs/store';

export const doneIssueSelector = IssueState.getIssuesByStatus(IssueStatus.DONE);
export const todoIssueSelector = IssueState.getIssuesByStatus(IssueStatus.TODO);
export const inProgressIssueSelector = IssueState.getIssuesByStatus(IssueStatus.IN_PROGRESS);

@Component({
  selector: 'app-board',
  templateUrl: './board.component.html',
  styleUrls: ['./board.component.scss']
})
export class BoardComponent {
  private readonly IssueStatus = IssueStatus;

  @Select(doneIssueSelector) doneIssues$: Observable<Issue[]>;
  @Select(todoIssueSelector) todoIssues$: Observable<Issue[]>;
  @Select(inProgressIssueSelector) inPRIssues$: Observable<Issue[]>;

  constructor() {
  }
}

