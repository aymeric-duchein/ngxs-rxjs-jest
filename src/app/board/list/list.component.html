<h3>{{ status }}</h3>
<div
  class="list"
  cdkDropList
  [cdkDropListData]="issues"
  (cdkDropListDropped)="drop($event)">
  <app-issue cdkDrag *ngFor="let issue of issues" [issue]="issue" [cdkDragData]="issue"
             (issueClick)="issueClick($event)"></app-issue>
</div>
