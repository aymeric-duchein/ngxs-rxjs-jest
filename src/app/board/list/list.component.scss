div {
  cursor: pointer;
  flex: auto;
  overflow: auto;

  &::-webkit-scrollbar {
    width: 5px;
    height: 8px;
    background: rgba(9, 30, 66, .08);
    border-radius: 3px;
  }

  &::-webkit-scrollbar-thumb {
    background-color: #bfc4ce;
    border-radius: 3px;
  }
}

h3 {
  margin-left: 1em;
}

:host {
  padding-bottom: 10px;
  padding-left: 3px;
  padding-right: 3px;
}

app-issue {
  cursor: pointer;
}

