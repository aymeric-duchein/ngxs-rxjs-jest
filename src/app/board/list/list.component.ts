import {Component, Input} from '@angular/core';
import {Issue, IssueStatus} from '../../core/issue/issue.model';
import {CdkDragDrop} from '@angular/cdk/drag-drop';
import {Store} from '@ngxs/store';
import {BoardIssueClick, BoardIssueDropped} from '../board.actions';

@Component({
  selector: 'app-list',
  templateUrl: './list.component.html',
  styleUrls: ['./list.component.scss']
})
export class ListComponent {

  @Input() status: IssueStatus;
  @Input() issues: Issue[];

  constructor(private readonly store: Store) {
  }

  drop(event: CdkDragDrop<Issue[], Issue[]>) {
    this.store.dispatch(new BoardIssueDropped(event.item.data, this.status));
  }

  issueClick(issue: Issue) {
    this.store.dispatch(new BoardIssueClick(issue));
  }
}
