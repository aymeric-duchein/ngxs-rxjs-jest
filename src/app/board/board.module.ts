import {NgModule} from '@angular/core';
import {CommonModule} from '@angular/common';
import {BoardComponent} from './board.component';
import {ListComponent} from './list/list.component';
import {DragDropModule} from '@angular/cdk/drag-drop';
import {IssueModule} from '../shared/issue/issue.module';
import {AddModule} from '../shared/add/add.module';

@NgModule({
  declarations: [BoardComponent, ListComponent],
  imports: [
    CommonModule,
    DragDropModule,
    IssueModule,
    AddModule,
  ]
})
export class BoardModule {
}
