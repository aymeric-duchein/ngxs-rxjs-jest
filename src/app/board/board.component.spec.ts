import {ComponentFixture, TestBed} from '@angular/core/testing';

import {BoardComponent, doneIssueSelector, inProgressIssueSelector, todoIssueSelector} from './board.component';
import {Issue} from '@src/app/core/issue/issue.model';
import {Subject} from 'rxjs';
import {mockSelect} from '@ngxs-labs/testing/jest';
import {MOCK_DONE_ISSUE, MOCK_IN_PROGRESS_ISSUE, MOCK_TODO_ISSUE} from '@mocks/issue.mock';
import {MockComponent, MockModule} from 'ng-mocks';
import {ListComponent} from '@src/app/board/list/list.component';
import {DragDropModule} from '@angular/cdk/drag-drop';
import {NgxsModule} from '@ngxs/store';
import {AddModule} from '@src/app/shared/add/add.module';

describe('BoardComponent', () => {
  let component: BoardComponent;
  let fixture: ComponentFixture<BoardComponent>;
  let doneIssueSelector$: Subject<Issue[]>;
  let todoIssueSelector$: Subject<Issue[]>;
  let inProgressIssueSelector$: Subject<Issue[]>;
  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [
        BoardComponent,
        MockComponent(ListComponent),
      ],
      imports: [
        NgxsModule.forRoot(),
        MockModule(DragDropModule),
        MockModule(AddModule)
      ]
    });

    doneIssueSelector$ = mockSelect(doneIssueSelector);
    todoIssueSelector$ = mockSelect(todoIssueSelector);
    inProgressIssueSelector$ = mockSelect(inProgressIssueSelector);
    fixture = TestBed.createComponent(BoardComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();

    doneIssueSelector$.next([MOCK_DONE_ISSUE]);
    todoIssueSelector$.next([MOCK_TODO_ISSUE]);
    inProgressIssueSelector$.next([MOCK_IN_PROGRESS_ISSUE]);

    fixture.detectChanges();
  });

  test('should create', () => {
    expect(component).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });
});
