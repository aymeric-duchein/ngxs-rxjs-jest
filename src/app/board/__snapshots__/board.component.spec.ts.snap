// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoardComponent should create 1`] = `
<app-board
  IssueStatus={[Function Object]}
  __doneIssues$__selector={[Function Subject]}
  __inPRIssues$__selector={[Function Subject]}
  __todoIssues$__selector={[Function Subject]}
>
  <div
    cdkdroplistgroup=""
    class="board-list-container"
  >
    <app-list
      class="mat-elevation-z2"
      ng-reflect-issues="[object Object]"
      ng-reflect-status="To do"
    />
    <app-list
      class="mat-elevation-z2"
      ng-reflect-issues="[object Object]"
      ng-reflect-status="In progress"
    />
    <app-list
      class="mat-elevation-z2"
      ng-reflect-issues="[object Object]"
      ng-reflect-status="Done"
    />
  </div><app-add />
</app-board>
`;
